x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/PrecompiledCharMapNormalizer.java

```diff
@@ -189,10 +189,20 @@ Reader normalize(CharSequence str) {
                 if (maybeSubStr.isPresent()) {
                     BytesRef subStr = maybeSubStr.get();
                     int numChars = UnicodeUtil.UTF8toUTF16(subStr.bytes, subStr.offset, subStr.length, reusableCharDecodeBuffer);
-                    normalizedCharPos += numChars;
-                    if (numChars != end - startIter) {
-                        addOffCorrectMap(normalizedCharPos, getLastCumulativeDiff() + end - startIter - numChars);
+                    int charDelta = numChars - (end - startIter);
+                    if (charDelta > 0) {
+                        // inserted chars, add the offset in the output stream
+                        int startOffset = normalizedCharPos + (end - startIter);
```

Comment:
this part slightly confuses me. Using `end - startIter` is the non-normalized number of chars. So, here you are saying:

 - take the last normalized position and then increase it by the non-normalized number of bytes
 - adjust the correction map by the diff

I am concerned about incrementing the offset correction values based on the non-normalized char positions

Maybe I am worrying about nothing. Its been awhile since I read this code. 

FYI, I have been mentally comparing this to: 

https://github.com/apache/lucene/blob/9af3ef8952597ead37a5a9ce5373ed3bbd40238e/lucene/analysis/common/src/java/org/apache/lucene/analysis/charfilter/MappingCharFilter.java#L138-L163